src/k2/httpproxy/HTTPProxy.cpp

```diff
@@ -186,13 +191,41 @@ seastar::future<> HTTPProxy::start() {
     return _startFut;
 }
 
+void HTTPProxy::TxnTracker::resetTimeout(Duration timeout) {
+    timer.rearm(Clock::now() + timeout);
+}
+
 seastar::future<nlohmann::json> HTTPProxy::_handleBegin(nlohmann::json&& request) {
     (void) request;
     return _client.beginTxn(k2::K2TxnOptions())
     .then([this] (auto&& txn) {
         K2LOG_D(k2::log::httpproxy, "begin txn: {}", txn.mtr());
-        _txns[_txnID++] = std::move(txn);
-        return JsonResponse(Statuses::S201_Created("Begin txn success"), nlohmann::json{{"txnID", _txnID - 1}});
+        auto txnid = _txnID++;
+        seastar::timer<> timer([this, txnid] {
```

Comment:
I don't think we should have a timer for every txn, they can be a bit expensive in seastar because changing their timeout time is a system call. We should have one timer that tracks the txn that is soonest to timeout. Their is an example of this in the module txn timeouts.

Also you should use the timers in src/k2/common/Timer.h which wraps the seastar timers and fixes some issues.

Reply:
Done.